feat: add fonts

jonowu opened the pull request:
## Purpose

Add fonts across the app.

## Changes

- Add font support
- Refactored the app to use 100% typography components.

### Screenshots
<img width="501" alt="Screen Shot 2020-10-17 at 6 18 04 pm" src="https://user-images.githubusercontent.com/53251695/96331019-8b70b600-10a5-11eb-8045-5fbe297b2f52.png">
<img width="501" alt="Screen Shot 2020-10-17 at 6 18 11 pm" src="https://user-images.githubusercontent.com/53251695/96331010-6bd98d80-10a5-11eb-80ce-bf22ce8fc8a0.png">
<img width="501" alt="Screen Shot 2020-10-17 at 6 18 36 pm" src="https://user-images.githubusercontent.com/53251695/96331013-71cf6e80-10a5-11eb-818d-a6cd19933704.png">
<img width="501" alt="Screen Shot 2020-10-17 at 6 18 39 pm" src="https://user-images.githubusercontent.com/53251695/96331015-73009b80-10a5-11eb-8d82-db560412b9be.png">
<img width="501" alt="Screen Shot 2020-10-17 at 6 18 43 pm" src="https://user-images.githubusercontent.com/53251695/96331016-73993200-10a5-11eb-8d67-d49af743788d.png">

## Checklist

- [ ] Tested on Android
- [ ] Tested on iOS

BenjaminYong commented on screens/action/index.js:
```diff
@@ -56,11 +56,7 @@ function ActionScreen({ route, navigation }) {
   return (
-    <ChildScreen
-      heading={title}
-      headerColor={colors.swinRed}
```
The red here added a red background on the child screen header if no images were provided on Strapi. Should we update the default headerColor prop in ChildScreen component to a yellow or red rather than gray?

BenjaminYong commented on components/typography/index.js:
```diff
@@ -26,8 +28,8 @@ const H5 = styled(H)`
 const Heading = styled(Core)`
-  color: ${({ color }) => (color ? color : colors.yellow)};
-  font-weight: bold;
+  color: ${({ primary }) => (primary ? colors.yellow : colors.black)};
```
<img src='https://user-images.githubusercontent.com/53241741/96331673-da6d1a00-10aa-11eb-9f53-60107898487c.png' width='25%' height='25%'/>

Heading typography component takes a 'primary' prop rather than 'color' as before.
Heading texts are rendering in black rather the provided color.

BenjaminYong replied:
Could update the Action Completion heading to pass thru primary prop

BenjaminYong replied:
The same applies to the submit-idea-completion screen, the 'Thank You!' there has gone black as it doesnt use the `primary `prop

BenjaminYong replied:
@jonowu Hey Jono, submit-idea-completion and action-completion headings havent been updated to use primary yet, rendering in black still

jonowu replied:
My bad, didn't push

BenjaminYong commented on screens/profile/index.js:
```diff
@@ -151,11 +151,15 @@ function ProfileScreen({ navigation, route }) {
-            {profileInfo.username && <Heading variant={3}>{profileInfo.username}</Heading>}
+            {profileInfo.username && (
+              <Heading primary variant={3}>
+                {profileInfo.username}
```
Should we make a transition to use the user's first name, like in the Submit Idea heading 'Hello {profileInfo.firstName}'?

BenjaminYong commented on screens/leaderboard/index.js:
```diff
@@ -107,22 +107,18 @@ function PodiumItem({ data, index }) {
-        style={{ width: '90%', textAlign: 'center', fontWeight: 'bold', fontSize: 19 }}
-      >
+      <Label adjustsFontSizeToFit numberOfLines={1} style={{ width: '90%', textAlign: 'center' }}>
         {data.name}
```
firstName update?

BenjaminYong commented on screens/leaderboard/index.js:
```diff
@@ -148,18 +144,13 @@ function LeaderboardItem({ data, index, current }) {
-            style={{ fontWeight: 'bold', marginLeft: 10, marginTop: 18 }}
-          >
+          <Label variant={2} adjustsFontSizeToFit numberOfLines={1} style={{ marginLeft: 10, marginTop: 18 }}>
             {data.name}
```
firstName update?

BenjaminYong commented on screens/feed/index.js:
```diff
@@ -249,9 +219,9 @@ function FeedScreen({ navigation }) {
                         />
-                        <Subheading variant={3} style={{ marginTop: 18 }}>
+                        <Heading variant={4} style={{ marginTop: 18 }}>
                           {entry.user.username}
```
firstName update?

BenjaminYong commented on screens/feed/index.js:
```diff
@@ -272,6 +242,40 @@ function FeedScreen({ navigation }) {
+                  {article.image ? (
+                    <TileImage source={{ uri: article.image.formats.small.url }}></TileImage>
+                  ) : (
+                    <TileAltImage></TileAltImage>
```
<img src='https://user-images.githubusercontent.com/53241741/96332105-fe7e2a80-10ad-11eb-8f34-710f325aa720.png' width='25%' height='25%'/>

This TileAltImage component doesn't style as expected with the latest changes, takes up half the screen's height

BenjaminYong replied:
This is the alternative header that appears if no image is provided for news articles

BenjaminYong replied:
When going into the article itself, it uses a swin red background color for the child screen header which renders fine - should sync up with the no image default for action headers

BenjaminYong reviewed (changes requested):
Nice stuff adding the fonts in, the app is starting to look really complete 😍 
Just a few comments around the `primary` prop for `Heading` typography and header styling for articles/actions without an image.
The username to first name transition could be done in a separate branch, but I reckon the change is justified to avoid inappropriate/personal usernames.

jonowu commented:
<img width="350" alt="Screen Shot 2020-10-18 at 1 00 12 pm" src="https://user-images.githubusercontent.com/53251695/96357040-ec4dcc00-1141-11eb-9fc6-55eb513e01f7.png">
<img width="350" alt="Screen Shot 2020-10-18 at 12 53 53 pm" src="https://user-images.githubusercontent.com/53251695/96357043-ef48bc80-1141-11eb-877c-06e017136cc0.png">
<img width="350" alt="Screen Shot 2020-10-18 at 12 54 03 pm" src="https://user-images.githubusercontent.com/53251695/96357047-f40d7080-1141-11eb-8ab8-b31e3c39df6a.png">
<img width="350" alt="Screen Shot 2020-10-18 at 12 59 51 pm" src="https://user-images.githubusercontent.com/53251695/96357059-199a7a00-1142-11eb-9177-685f59b60d64.png">

jonowu commented:
Cheers for the thorough review, I've addressed your comments @BenjaminYong. Lodash has a nice `capitalize()` function which I'm using to make sure the first names have a capital letter 😁

BenjaminYong commented on screens/leaderboard/index.js:
```diff
@@ -205,9 +195,9 @@ function LeaderboardScreen() {
-    const isUserOutsideTop10 = !top10Users.some((user) => user.name === username);
-    const currentUserData = leaderboard.find((user) => user.name === username);
-    const currentUserRanking = leaderboard.findIndex((user) => user.name === username);
+    const isUserOutsideTop10 = !top10Users.some((user) => user.id === userId);
```
Not sure if this is working: just made a new user, completed an action, visualisation is updated (good), but leader board doesn't show their placing outside the top 10. Rebuilt the app, but they're still not in the leader board

BenjaminYong replied:
Just noticed they the current user only appears once they are within the top 10, and also the current user (Constantine-Elizabeth) does not highlight in yellow anymore. 
<img src='https://user-images.githubusercontent.com/53241741/96357330-2ae58580-1146-11eb-8cb3-a08bca4ea88b.png' width='25%' height='25%'/>

BenjaminYong commented on screens/profile/index.js:
```diff
@@ -151,11 +151,15 @@ function ProfileScreen({ navigation, route }) {
-            {profileInfo.username && <Heading variant={3}>{profileInfo.username}</Heading>}
+            {profileInfo.username && (
+              <Heading primary variant={3}>
+                {`${_.capitalize(profileInfo.firstName)} ${_.capitalize(profileInfo.lastName)}`}
```
What a sweet function! Just only noticed that names become off-centre when they are too long
<img src='https://user-images.githubusercontent.com/53241741/96357378-d5f63f00-1146-11eb-9a5b-fa967414acb4.png' width='25%' height='25%'/>

jonowu replied:
Fixed

BenjaminYong commented on screens/feed/index.js:
```diff
@@ -249,9 +220,9 @@ function FeedScreen({ navigation }) {
-                          {entry.user.username}
-                        </Subheading>
+                        <Heading variant={4} style={{ marginTop: 18 }}>
+                          {capitalize(entry.user.firstName)}
```
Similar case here, where the name clips into the right text element

<img src='https://user-images.githubusercontent.com/53241741/96357426-4b620f80-1147-11eb-91e2-a8375f63cc25.png' height='25%' width='25%'/>

jonowu replied:
Fixed

BenjaminYong reviewed (changes requested):
Thanks for making those updates @jonowu 😀 
Found a few other tweaks:
- Leaderboard doesn't show current logged in user's ranking if outside top 10
- Leaderboard doesn't highlight the current logged in user in yellow anymore
- Long names clip into leaderboard's number of actions completed text
- Long names become off centre in the profile heading
- Long names clip into the social feed's 'x time ago' text

jonowu commented:
Cheers @BenjaminYong, that should be everything fixed now.

Leaderboard highlighting wasn't working because I switch to using the IDs instead of username for comparison, but the IDs were coming in as strings from strapi. Converting them to ints fixed it.

Also included a fun colour change in the leaderboard ;)

BenjaminYong reviewed (approved):
LGBT! 👏🏼 
Love the Olympic medals 🥇🥈🥉  - omg they are emojis 😮
